File: latest/week3/workshop7_m2/W07_M02_demo.py
"""W07-M02: Exploratory vs Explanatory Visualization — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.dates import DateFormatter
os.makedirs("output", exist_ok=True)

# ── 1. LOAD DATA ────────────────────────────────────────
nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf["year_added"] = nf["date_added"].dt.year
yearly = (nf.query("2015 <= year_added <= 2021")
    .groupby(["year_added", "type"]).size().reset_index(name="n"))
movies = yearly.query("type == 'Movie'")
tvshows = yearly.query("type == 'TV Show'")

# ── 2. READER-DRIVEN vs AUTHOR-DRIVEN ───────────────────
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(16, 5.5))

# Reader-driven (dashboard)
x = np.arange(2015, 2022); w = 0.3
for t, offset, c in [("Movie", -w/2, "#1565C0"), ("TV Show", w/2, "#E53935")]:
    sub = yearly.query("type == @t")
    ax1.bar(sub.year_added + offset, sub.n, width=w, color=c, alpha=0.7, label=t)
ax1.legend(fontsize=8)
ax1.set_title("READER-DRIVEN (Dashboard)\n'Netflix Additions by Type'\n"
    "Descriptive title, legend, both series equal weight",
    fontsize=10, fontweight="bold", color="#888")
ax1.set_xlabel("Year"); ax1.set_ylabel("Titles")
ax1.grid(axis="y", alpha=0.2)

# Author-driven (story slide)
ax2.plot(tvshows.year_added, tvshows.n, "-o", color="#DDDDDD", lw=0.8, markersize=3)
ax2.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2.5, markersize=6)
ax2.text(2021.15, movies.iloc[-1]["n"], " Movies", fontsize=9,
    fontweight="bold", color="#1565C0", va="center")
ax2.text(2021.15, tvshows.iloc[-1]["n"], " TV Shows", fontsize=8,
    color="#BBBBBB", va="center")
# Decline callout
peak_n = movies["n"].max()
latest_n = movies.iloc[-1]["n"]
decline = round((1 - latest_n / peak_n) * 100)
ax2.annotate(f"–{decline}% since\n2019 peak", xy=(2020, (peak_n + latest_n)/2),
    fontsize=10, fontweight="bold", color="#E53935", ha="center",
    bbox=dict(boxstyle="round,pad=0.3", facecolor="white", edgecolor="#E53935", alpha=0.9))
ax2.set_xlim(2014.5, 2022.5)
ax2.set_title("AUTHOR-DRIVEN (Story Slide)\n"
    f"'Movie additions declined {decline}% from 2019 peak'\n"
    "Declarative title, grey+accent, direct labels",
    fontsize=10, fontweight="bold", color="#1565C0")
ax2.set_ylabel("Titles Added")

for ax in [ax1, ax2]:
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
fig.suptitle("Same Data: Reader-Driven (Dashboard) vs Author-Driven (Story Slide)",
    fontsize=13, fontweight="bold")
plt.tight_layout(); plt.savefig("output/reader_vs_author.png", dpi=300); plt.close()

# ── 3. MARTINI GLASS: 3 Slides → Dashboard ──────────────
fig, axes = plt.subplots(2, 2, figsize=(14, 9))

# Slide 1: Context
cumul = nf.dropna(subset=["date_added"]).set_index("date_added").resample("ME").size().cumsum()
axes[0,0].fill_between(cumul.index, cumul.values, alpha=0.15, color="#1565C0")
axes[0,0].plot(cumul.index, cumul.values, color="#1565C0", lw=1.2)
axes[0,0].set_title("Slide 1: Netflix built 8,800+ titles",
    fontsize=10, fontweight="bold", color="#2E7D32")
axes[0,0].xaxis.set_major_formatter(DateFormatter("%Y"))

# Slide 2: Complication
axes[0,1].plot(tvshows.year_added, tvshows.n, "-o", color="#DDD", lw=0.8, markersize=3)
axes[0,1].plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2, markersize=5)
axes[0,1].set_title("Slide 2: But movies declining since 2019",
    fontsize=10, fontweight="bold", color="#E65100")

# Slide 3: Resolution
nf["primary_country"] = nf["country"].str.split(",").str[0].str.strip()
top5 = nf.dropna(subset=["primary_country"])["primary_country"].value_counts().head(5)
axes[1,0].barh(top5.index[::-1], top5.values[::-1], color="#2E7D32", height=0.5)
axes[1,0].set_title("Slide 3: US leads, India rising",
    fontsize=10, fontweight="bold", color="#7B1FA2")

# Dashboard transition
monthly = (nf.dropna(subset=["date_added"]).query("date_added.dt.year >= 2015")
    .set_index("date_added").groupby("type").resample("ME").size()
    .reset_index(name="n"))
for t, c in [("Movie", "#1565C0"), ("TV Show", "#E53935")]:
    sub = monthly.query("type == @t")
    axes[1,1].plot(sub["date_added"], sub["n"], color=c, lw=0.5, label=t)
axes[1,1].legend(fontsize=6)
axes[1,1].set_title("→ DASHBOARD: Explore freely (filter, drill down)",
    fontsize=10, fontweight="bold", color="#888")
axes[1,1].xaxis.set_major_formatter(DateFormatter("%Y"))

for ax in axes.flatten():
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
    ax.tick_params(labelsize=6)
fig.suptitle("Martini Glass: 3 Guided Slides → Dashboard Exploration",
    fontsize=13, fontweight="bold")
plt.tight_layout(); plt.savefig("output/martini_glass.png", dpi=300); plt.close()

# ── 4. e-Car: DASHBOARD vs STORY ────────────────────────
ec = pd.read_csv("ecar.csv"); ec.columns = [c.strip().replace("  "," ") for c in ec.columns]
ec["Year"] = pd.to_datetime(ec["Approve Date"], format="%m/%d/%Y", errors="coerce").dt.year
ec["Spread"] = ec["Rate"] - ec["Cost of Funds"]
yr_ec = ec.query("2002<=Year<=2012").groupby("Year").agg(
    rate=("Rate","mean"), spread=("Spread","mean")).reset_index()

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 5))
# Dashboard
ax1.plot(yr_ec.Year, yr_ec.rate, "-o", color="#1565C0", lw=1.2, label="Rate")
ax1.plot(yr_ec.Year, yr_ec.spread, "-s", color="#E53935", lw=1.2, label="Spread")
ax1.legend(fontsize=7)
ax1.set_title("DASHBOARD: 'Rate and Spread Over Time'\n(descriptive, two equal series)",
    fontsize=9, fontweight="bold", color="#888")
# Story
ax2.plot(yr_ec.Year, yr_ec.spread, color="#DDD", lw=0.8)
ax2.plot(yr_ec.Year, yr_ec.rate, "-o", color="#1565C0", lw=2.5, markersize=5)
ax2.axvline(x=2008, ls="--", color="#E53935", lw=0.8)
ax2.annotate("2008 crisis:\n–2pp rate drop", xy=(2008, yr_ec.rate.min()),
    xytext=(2009.5, yr_ec.rate.max()*0.85),
    fontsize=8, fontweight="bold", color="#E53935",
    arrowprops=dict(arrowstyle="->", color="#E53935", lw=1))
ax2.text(2012.1, yr_ec.iloc[-1].rate, " Rate", fontsize=8,
    fontweight="bold", color="#1565C0", va="center")
ax2.text(2012.1, yr_ec.iloc[-1].spread, " Spread", fontsize=7,
    color="#BBB", va="center")
ax2.set_xlim(2001.5, 2013.5)
ax2.set_title("STORY: 'Loan rates dropped 2pp after the 2008 crisis'\n(declarative, grey+accent, annotated)",
    fontsize=9, fontweight="bold", color="#1565C0")
for ax in [ax1,ax2]:
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
    ax.set_ylabel("Percentage Points")
fig.suptitle("e-Car: Same Data, Dashboard vs Story Slide", fontweight="bold")
plt.tight_layout(); plt.savefig("output/ecar_dash_vs_story.png", dpi=300); plt.close()

# ── 5. TRANSFORMATION CHECKLIST ──────────────────────────
print("\n=== Dashboard → Story Slide Checklist ===")
checks = [
    ("Title", "Descriptive → Declarative (the finding)"),
    ("Legend", "Remove → Direct labels at endpoints"),
    ("Colour", "All coloured → Grey + accent"),
    ("Gridlines", "Visible → Minimal or none"),
    ("Annotation", "None → Callouts, arrows, % labels"),
    ("Data density", "Show all → Show only the story"),
    ("Interactivity", "Filters/tooltips → Static"),
    ("Subtitle", "Metric description → Strategic implication"),
]
for element, change in checks:
    print(f"  {element:<15} {change}")

print("\nAll W07-M02 Python outputs saved")

File: latest/week3/workshop7_m2/W07_M02_demo.R
# ============================================================
# Workshop 7 — Module 2: Exploratory vs Explanatory Visualization
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output", showWarnings = FALSE)

# ── 1. LOAD DATA ────────────────────────────────────────
nf <- read_csv("netflix.csv") |>
  mutate(date_added = mdy(str_trim(date_added)),
    year_added = year(date_added),
    primary_country = str_trim(str_extract(country, "^[^,]+")))

yearly_type <- nf |>
  filter(!is.na(year_added), year_added >= 2015, year_added <= 2021) |>
  count(year_added, type)

# ── 2. READER-DRIVEN (Dashboard-Style) ──────────────────
# Grouped bar chart: both types visible, legend, gridlines, descriptive title
p_dashboard <- ggplot(yearly_type, aes(x = factor(year_added), y = n, fill = type)) +
  geom_col(position = "dodge", width = 0.7) +
  scale_fill_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  theme_minimal() +
  theme(panel.grid.major.x = element_blank()) +
  labs(title = "Netflix Additions by Type (2015–2021)",
       subtitle = "Dashboard view: user reads both series, draws own conclusions",
       x = "Year", y = "Titles Added", fill = "Type")
ggsave("output/reader_driven.png", p_dashboard, width = 9, height = 5, dpi = 300)

# ── 3. AUTHOR-DRIVEN (Story Slide) ──────────────────────
# Grey+accent, declarative title, direct labels, annotation
p_story <- ggplot(yearly_type, aes(x = year_added, y = n, group = type)) +
  geom_line(color = "#DDDDDD", linewidth = 0.8) +
  geom_line(data = yearly_type |> filter(type == "Movie"),
    color = "#1565C0", linewidth = 2.5) +
  geom_point(data = yearly_type |> filter(type == "Movie"),
    color = "#1565C0", size = 3) +
  # Direct labels
  geom_text(data = yearly_type |> filter(year_added == 2021),
    aes(label = type, color = type == "Movie"),
    hjust = -0.15, size = 3.5, fontface = "bold", show.legend = FALSE) +
  scale_color_manual(values = c("TRUE" = "#1565C0", "FALSE" = "#BBBBBB")) +
  # Decline callout
  annotate("segment", x = 2019, xend = 2021, y = 1050, yend = 1050,
    linetype = "dotted", color = "#888") +
  annotate("text", x = 2020, y = 950,
    label = "–42% since\n2019 peak", size = 3.5, fontface = "bold",
    color = "#E53935") +
  coord_cartesian(xlim = c(2015, 2022.5)) +
  theme_minimal(base_size = 11) +
  theme(legend.position = "none",
    plot.title = element_text(face = "bold", size = 13)) +
  labs(title = "Movie additions declined 42% from their 2019 peak",
       subtitle = "Author-driven: one message, annotated, no legend",
       x = NULL, y = "Titles Added per Year")
ggsave("output/author_driven.png", p_story, width = 9, height = 5, dpi = 300)

# Side-by-side comparison
ggsave("output/reader_vs_author.png",
  (p_dashboard + labs(title = "READER-DRIVEN\n(Dashboard: descriptive, interactive)")) |
  (p_story + labs(title = "AUTHOR-DRIVEN\n(Story slide: declarative, focused)")),
  width = 16, height = 5.5, dpi = 300)

# ── 4. MARTINI GLASS: Narrative → Dashboard ──────────────
# Part 1: 3-slide author-driven narrative
p_slide1 <- ggplot(nf |> filter(!is.na(date_added)) |>
  mutate(month = floor_date(date_added, "month")) |>
  count(month) |> mutate(cum = cumsum(n)),
  aes(x = month, y = cum)) +
  geom_area(fill = "#1565C0", alpha = 0.15) +
  geom_line(color = "#1565C0", linewidth = 1) +
  theme_minimal(base_size = 8) +
  labs(title = "Slide 1: Netflix built 8,800+ titles by 2021",
       x = NULL, y = "Cumulative")

p_slide2 <- p_story +
  labs(title = "Slide 2: But movies are declining since 2019") +
  theme(plot.title = element_text(size = 9))

top5_countries <- nf |> filter(!is.na(primary_country)) |>
  count(primary_country, sort = TRUE) |> slice_max(n, n = 5)
p_slide3 <- ggplot(top5_countries,
  aes(x = fct_reorder(primary_country, n), y = n)) +
  geom_col(fill = "#2E7D32", width = 0.5) +
  geom_text(aes(label = format(n, big.mark = ",")),
    hjust = -0.1, size = 2.5, fontface = "bold") +
  coord_flip(clip = "off") +
  theme_minimal(base_size = 8) +
  labs(title = "Slide 3: US dominates, but India is rising fast",
       x = NULL, y = "Total Titles")

# Part 2: dashboard (reader-driven)
monthly_all <- nf |>
  filter(!is.na(date_added), year(date_added) >= 2015) |>
  mutate(month = floor_date(date_added, "month")) |>
  count(month, type)
p_dash <- ggplot(monthly_all, aes(x = month, y = n, color = type)) +
  geom_line(linewidth = 0.5) +
  scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  theme_minimal(base_size = 7) +
  labs(title = "TRANSITION → DASHBOARD (explore freely)",
       subtitle = "Filter by type, country, genre, date range",
       x = NULL, y = "Monthly Additions", color = "Type")

ggsave("output/martini_glass.png",
  (p_slide1 | p_slide2 | p_slide3) / p_dash +
  plot_annotation(
    title = "Martini Glass: 3 Guided Slides → Dashboard Exploration",
    tag_levels = list(c("Slide 1", "Slide 2", "Slide 3", "Dashboard"))),
  width = 14, height = 9, dpi = 300)

# ── 5. e-Car: Dashboard vs Story ────────────────────────
ec <- read_csv("ecar.csv") |>
  mutate(Year = year(mdy(`Approve Date`)),
    Spread = Rate - `Cost of Funds`) |>
  filter(Year >= 2002, Year <= 2012)

yearly_ec <- ec |> group_by(Year) |>
  summarise(rate = mean(Rate), spread = mean(Spread), n = n())

# Dashboard version
p_ec_dash <- ggplot(yearly_ec) +
  geom_line(aes(x = Year, y = rate), color = "#1565C0", linewidth = 1) +
  geom_line(aes(x = Year, y = spread), color = "#E53935", linewidth = 1) +
  theme_minimal(base_size = 9) +
  labs(title = "e-Car: Rate and Spread (Dashboard)",
       subtitle = "Descriptive — user interprets both lines",
       y = "pp")

# Story slide version
p_ec_story <- ggplot(yearly_ec) +
  geom_line(aes(x = Year, y = spread), color = "#DDDDDD", linewidth = 0.8) +
  geom_line(aes(x = Year, y = rate), color = "#1565C0", linewidth = 2.5) +
  geom_vline(xintercept = 2008, linetype = "dashed", color = "#E53935") +
  annotate("text", x = 2008.3, y = max(yearly_ec$rate) * 0.9,
    label = "2008\ncrisis", color = "#E53935", size = 3, fontface = "bold") +
  theme_minimal(base_size = 9) +
  theme(legend.position = "none") +
  labs(title = "Loan rates dropped 2pp after the 2008 crisis",
       subtitle = "Author-driven — one message: the crisis impact on rates",
       y = "Rate (pp)")

ggsave("output/ecar_dash_vs_story.png",
  p_ec_dash | p_ec_story, width = 14, height = 5, dpi = 300)

# ── 6. TRANSFORMATION CHECKLIST (printed) ────────────────
cat("\n── Transformation Checklist: Dashboard → Story Slide ──\n")
cat("1. Title: Descriptive → Declarative (state the finding)\n")
cat("2. Legend: Remove → Direct labels at line endpoints\n")
cat("3. Colour: All coloured → Grey+accent (one story line)\n")
cat("4. Gridlines: Visible → Minimal or none\n")
cat("5. Annotations: None → Callouts, arrows, % change labels\n")
cat("6. Data density: Show all → Show only the story\n")
cat("7. Interactivity: Filters/tooltips → None (static)\n")
cat("8. Subtitle: Metric description → Strategic implication\n")

cat("\n── All W07-M02 R outputs saved ──\n")
